src/blockstore/log.go

```diff
@@ -0,0 +1,443 @@
+	}
+
+	for i := 0; i < count; i++ {
+		kv = new(keyVal)
+		if kv == nil {
+			return nil, errors.New("Could not allocate memory.")
+		}
+		tmp = binary.LittleEndian.Uint64(src[i*8 : i*8+8])
+		kv.key, kv.flags = decodeKeyFlags(tmp)
+		dst[i] = kv
+	}
+	return dst, nil
+
+}
+
+// A helper function that generates the header for the log file to be written.
+func generateHeader(lh *logHeader) ([]byte, error) {
+	var headerSize int = defaultHeaderSize
+	res := make([]byte, headerSize, headerSize)
+
+	if res == nil {
+		log.Println("Could not allocate memory for header.")
+		return nil, errors.New("Could not allocate memory for header.")
+	}
+	// Writing the version of the header.
+	binary.LittleEndian.PutUint64(res[0:8], uint64(lh.headerVersion))
+
+	// Writing the size of header itself.
+	binary.LittleEndian.PutUint64(res[8:16], uint64(headerSize))
+
+	// Writing the size of indexTable.
+	binary.LittleEndian.PutUint64(res[16:24], uint64(lh.numOfKeys))
+
+	//TODO adding merkleRoot and root of next log
+
+	return res, nil
+}
+
+/////////////////////////////////////////////////////
+// Methods that are called on a logManager object. //
+/////////////////////////////////////////////////////
+
+// A helper function that generates a file name.
+func (lm *logManager) generateName() string {
+	return lm.namePrefix + strconv.Itoa(lm.nextFD) + defaultSuffix
+}
+
+// Writes a log file at the head.
+// The input is a slice of keyVals that MUST be sorted by key.
+// During write, the flags are appended to key (1 byte flags + 7 bytes key)
+// TODO making sure to take care of removals.
+// TODO make sure it moves the head.
+func (lm *logManager) write(kvList []*keyVal) error {
+	var err error
+	var newHeader *logHeader
+	var headerBytes, indexBytes []byte
+	var bytesWritten int
+	var encryptedBlock *Block
+	var hashList [][]byte
+	var numOfLeaves, numOfNodes, firstLeafIndex int
+	var treeOffset int64
+
+	// Fist assign a new header file.
+	newHeader = new(logHeader)
+	if newHeader == nil {
+		log.Println("Could not allocate header.")
+		return errors.New("Could not allocate header.")
+	}
+	newHeader.headerVersion = defaultHeaderVersion
+	newHeader.numOfKeys = len(kvList)
+	if lm.headLog != nil {
+		newHeader.nextLog = lm.headLog
+		newHeader.nextRoot = lm.headLog.merkleRoot
+	}
+
+	//TODO move down after completion/verification of write to
+	// prevent corruption of head.
+	lm.headLog = newHeader
+
+	// Making sure we are writing a NEW file.
+	for checkFileExists(lm.generateName()) {
+		lm.nextFD++
+	}
+	newHeader.file, err = os.Create(lm.generateName())
+	if err != nil {
+		log.Println("Could not create log file.")
+		return errors.New("Could not create log file.")
+	}
+	lm.nextFD++
+
+	// Generate the header, key index table, and merkleTree in memory.
+	headerBytes, err = generateHeader(newHeader)
+	if err != nil {
+		return errors.New("Could not generate header.")
+	}
+
+	indexBytes, err = generateIndex(kvList)
```

Comment:
If the input is not sorted correctly, generateIndex will return an error. I suggest you propagate that error up, rather than throw it out and generate a new (less useful) error. In general it's probably good practice to propagate errors, rather than recreate different ones: please change this function to propagate existing errors rather than create new ones when possible

Reply:
Makes sense.

Reply:
The issue is propagating the error up, will not show the trace.

Reply:
I tried to make it better for now.